source/slang/slang-ir-fuse-satcoop.cpp

```diff
@@ -0,0 +1,513 @@
+    const auto newBroadcast = usesSameSharedInput
+        ? callF.broadcast
+        : makeBroadcastBoth(builder, callF.sharedInputType, callG.sharedInputType, callF.broadcast, callG.broadcast);
+    const auto newSharedInput = usesSameSharedInput
+        ? callF.sharedInput
+        : builder.emitMakeTuple(callF.sharedInput, callG.sharedInput);
+    const auto newDistinctInput = usesSameDistinctInput 
+        ? callF.distinctInput 
+        : builder.emitMakeTuple(callF.distinctInput, callG.distinctInput);
+
+    // Call it and extract the results from f and g
+    const auto res = builder.emitCallInst(
+        newRetType,
+        newSpec,
+        {newCooperate, newFallback, newWaveMatch, newBroadcast, newSharedInput, newDistinctInput});
+    const auto resF = builder.emitGetTupleElement(callF.retType, res, 0);
+    const auto resG = builder.emitGetTupleElement(callG.retType, res, 1);
+    f->replaceUsesWith(resF);
+    g->replaceUsesWith(resG);
+    f->removeAndDeallocate();
+    g->removeAndDeallocate();
+
+    return res;
+}
+
+//
+// Identify calls which we can fuse
+//
+IRCall* isCallNamed(const char* n, IRInst* i)
+{
+    auto call = as<IRCall>(i);
+    if(!call)
+        return nullptr;
+    // saturated_cooperation is a generic function, so look for specializations thereof
+    auto spec = as<IRSpecialize>(call->getCallee());
+    if(!spec)
+        return nullptr;
+    auto func = findSpecializeReturnVal(spec);
+    if(!func)
+        return nullptr;
+
+    // TODO: Is there nothing better for finding a specific declaration in
+    // core.slang?
+    auto h = func->findDecoration<IRNameHintDecoration>();
+    if(h->getName() != n)
+        return nullptr;
+    return call;
+}
+
+static void fuseCallsInBlock(IRBuilder& builder, IRBlock* block)
+{
+    // first, inline calls to saturated_cooperation to expose
+    // saturated_cooperation_using which is simpler to fuse.
+    // It is simpler to fuse because it makes explicit the inter-lane
+    // communication functions, which we can use as buiding blocks in our
+    // composition.
+
+    List<IRCall*> toInline;
+    for (auto inst : block->getChildren())
+    {
+        if(auto sat_coop = isCallNamed("saturated_cooperation", inst))
+            toInline.add(sat_coop);
+    }
+    for(auto c : toInline)
+        inlineCall(c);
+
+    // Walk over the instructions in this block
+    // If we see a call to sat_coop then remember where it is and keep
+    // walking, if we reach another call without first encountering any
+    // instructions with which our first call can't be safely reordered
+    // then we remove the first call and replace the second with a fused
+    // call.
+    IRCall* lastCall = nullptr;
+    for(auto inst = block->getFirstInst(); inst != block->getTerminator(); inst = inst->getNextInst())
+    {
+        if(auto call = isCallNamed("saturated_cooperation_using", inst))
```

Comment:
This name based matching does not seem to be very robust. I think it maybe better to use some special decoration to identify it.

Reply:
:100: